shared/libraries/config_protocol/include/config_protocol/config_protocol_client.h

```diff
@@ -186,29 +190,50 @@ ConfigProtocolClient<TRootDeviceImpl>::ConfigProtocolClient(const ContextPtr& da
 }
 
 template<class TRootDeviceImpl>
-void ConfigProtocolClient<TRootDeviceImpl>::protocolHandshake()
+void ConfigProtocolClient<TRootDeviceImpl>::protocolHandshake(uint16_t protocolVersion)
 {
     auto getProtocolInfoRequestPacketBuffer = PacketBuffer::createGetProtocolInfoRequest(clientComm->generateId());
     const auto getProtocolInfoReplyPacketBuffer = sendRequestCallback(getProtocolInfoRequestPacketBuffer);
 
+    const std::set<uint16_t> supportedClientVersions {0, 1};
+
     uint16_t currentVersion;
-    std::vector<uint16_t> supportedVersions;
-    getProtocolInfoReplyPacketBuffer.parseProtocolInfoReply(currentVersion, supportedVersions);
+    std::set<uint16_t> supportedServerVersions;
+    getProtocolInfoReplyPacketBuffer.parseProtocolInfoReply(currentVersion, supportedServerVersions);
 
-    if (currentVersion != 0)
-        throw ConfigProtocolException("Invalid server protocol version");
+    if (protocolVersion != std::numeric_limits<uint16_t>::max())
+    {
+        if (std::find(supportedClientVersions.begin(), supportedClientVersions.end(), protocolVersion) == supportedClientVersions.end())
+            throw ConfigProtocolException("Protocol not supported on client");
 
-    if (std::find(supportedVersions.begin(), supportedVersions.end(), 0) == supportedVersions.end())
-        throw ConfigProtocolException("Protocol not supported on server");
+        if (std::find(supportedServerVersions.begin(), supportedServerVersions.end(), protocolVersion) == supportedServerVersions.end())
+            throw ConfigProtocolException("Protocol not supported on server");
+    }
+    else
+    {
+        std::set<uint16_t> commonVersions;
+        std::set_intersection(supportedClientVersions.begin(),
+                              supportedClientVersions.end(),
+                              supportedServerVersions.begin(),
+                              supportedServerVersions.end(),
+                              std::inserter(commonVersions, commonVersions.begin()));
+
+        if (commonVersions.empty())
+            throw ConfigProtocolException("Cannot handshake a common protocol version");
 
-    auto upgradeProtocolRequestPacketBuffer = PacketBuffer::createUpgradeProtocolRequest(clientComm->generateId(), 0);
+        protocolVersion = *commonVersions.rbegin();
+    }
+
+    auto upgradeProtocolRequestPacketBuffer = PacketBuffer::createUpgradeProtocolRequest(clientComm->generateId(), protocolVersion);
     const auto upgradeProtocolReplyPacketBuffer = sendRequestCallback(upgradeProtocolRequestPacketBuffer);
 
     bool success;
     upgradeProtocolReplyPacketBuffer.parseProtocolUpgradeReply(success);
 
     if (!success)
         throw ConfigProtocolException("Protocol upgrade failed");
+
+    clientComm->setProtocolVersion(protocolVersion);
```

Comment:
Some logging would be useful here to aid in debugging what protocol version was handshaken.

Reply:
Ok